### PR TITLE
packetbeat: Enable setting promiscuous mode automatically 

### PR DESCRIPTION
**Pull request by michalpristas**
This PR tries to enable promiscuous mode when using `af_packet`.

Few things to mention
- when it fails it does not panic just logs the failure, mostly because of and backward compatibility
- on exit: it resets promisc mode to whatever it was before
- when `any` is used it does not set the mode to promiscuous - following `pcap` 

Fixes elastic/beats#700 

### Basic check:
- check promisc mode is disabled on your device (e.g. `ip link show {device}`) (look for `PROMISC` keyword)
- enable promisc mode by including `packetbeat.interfaces.auto_promisc_mode: true` into packetbeat config
- check promisc mode is enabled (e.g. `ip link show {device}`)
- stop packetbeat
- check promisc mode is disabled (e.g. `ip link show {device}`)

### Already on promisc check:
- enable promisc mode on your devide (e.g. `ifconfig {device} promisc` or `ip link set {device} promisc on`
- enable promisc mode by including `packetbeat.interfaces.auto_promisc_mode: true` into packetbeat config
- check promisc mode is enabled (e.g. `ip link show {device}`)
- stop packetbeat
- check promisc mode is still enabled (e.g. `ip link show {device}`)

### CODE REVIEW

**Comment by andrewkroh** [comment]
Thanks for looking into this issue.

I definitely don't want to disable the seccomp protections to get this feature. Is there a way to directly set the promisc mode? How does this `ip` command do it? Maybe run that command in `strace` to see what it does. It might be using `ioctl` with `IFF_PROMISC`. Or maybe it's some kind of netlink msg being sent to the kernel.

**Comment by michalpristas** [comment]
i believe it will definitely call ioctl. i will take a closer look at it and come with a possible workaround

**Comment by michalpristas** [comment]
@andrewkroh changed it to call SIOCGIFFLAGS syscall directly. tested with Ubuntu and seems to work

File: packetbeat/sniffer/afpacket_linux.go [hunk 1]
```diff
@@ -69,4 +87,42 @@ func (h *afpacketHandle) LinkType() layers.LinkType {
...
+	return ifl.flags&uint16(syscall.IFF_PROMISC) != 0, nil
+}
+
+func setPromiscMode(device string, enabled bool) error {
+	if device == "any" {
+		logp.Warn("Cannot set promiscuous mode to device 'any'")
+		return nil
+	}
+
+	return syscall.SetLsfPromisc(device, enabled)
```
Comment (andrewkroh) [inline packetbeat/sniffer/afpacket_linux.go]: This function says its deprecated. Have you checked out what it recommends to use?

Comment (michalpristas) [reply packetbeat/sniffer/afpacket_linux.go]: for now `x/net` where I did not find anything useful. so, for now, I went with this with a possible rewrite to direct syscalls

**Comment by michalpristas** [comment]
@andrewkroh I played with system tests a bit and added one which tests exactly the proposed scenario. the drawback is that it needs root access and while it runs locally on jenkins it is skipped (unless you know how to get NET_ADMIN)

**Comment by andrewkroh** [comment]
Thanks for adding the test case! You might be able to set TEST_ENVIRONMENT=true to have the tests executed inside of Docker. I think that requires the addition of a Dockerfile.

https://github.com/elastic/beats/blob/4ec5d138e403a363f893b08c3534860c757cfd2e/packetbeat/Makefile#L4

**Comment by michalpristas** [comment]
thanks @andrewkroh will try that

File: packetbeat/sniffer/afpacket_linux.go [hunk 2]
```diff
@@ -20,22 +20,42 @@
...
+		logp.Err("Failed to get promiscuous mode for device '%s': %v", device, err)
+	}
+
+	h := &afpacketHandle{
+		promiscPreviousState:         promiscEnabled,
+		device:                       device,
+		promiscPreviousStateDetected: err == nil,
+	}
+
+	if err := setPromiscMode(device, true); err != nil {
```
Comment (urso) [inline packetbeat/sniffer/afpacket_linux.go]: why change mode if it's already enabled?

File: packetbeat/sniffer/afpacket_linux.go (hunk 1 above)
Comment (urso) [reply packetbeat/sniffer/afpacket_linux.go]: Let's at least add a code comment here explaining the issue + follow up issue in repo.

**Comment by michalpristas** [comment]
👍 on doing this optional

**Comment by andrewkroh** [comment]
@michalpristas This feature is still on my wish list. 🙏 😄

**Comment by urso** [comment]
@michalpristas @andrewkroh We want to continue with this PR? What is missing?

**Comment by michalpristas** [comment]
@urso brought even with master and updated to answer your comments. do you think something is missing?

**Pull request merged**